Collect real time replication metrics for prometheus.

poornas opened the pull request:
Also adding pending_count metric for total pending
replication operations.

## Description


## Motivation and Context
Data usage info from scanner can lag behind - calculate inline metrics & use the most up-to-date values.

## How to test this PR?


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

Commit:
Collect real time replication metrics for prometheus.

Also adding pending_count metric for total pending
replication operations.

harshavardhana commented on cmd/bucket-replication-stats.go:
```diff
@@ -0,0 +1,142 @@
+type ReplicationStats struct {
+	Cache map[string]*BucketReplicationStats
+	sync.RWMutex
+}
```
```suggestion
type ReplicationStats struct {
	sync.RWMutex
	Cache map[string]*BucketReplicationStats
}
```

harshavardhana commented on cmd/bucket-replication.go:
```diff
@@ -851,8 +852,10 @@ func (p *ReplicationPool) queueReplicaTask(oi ObjectInfo) {
+	case <-GlobalContext.Done():
+		return
 	default:
+		p.replicaCh <- oi
```
This can block @poornas why did you change it like this? and it can hold up a go-routine we can't have that.

poornas replied:
if server receives Ctrl-C, there is a panic here if replication is in progress. We close the channels right away after global context is cancelled.

harshavardhana replied:
Yes but it will block if the channel is full @poornas 

We have to keep both in select and default should be to ignore the channel blockage..

harshavardhana commented on cmd/data-usage-cache.go:
```diff
@@ -46,14 +46,15 @@ type sizeHistogram [dataUsageBucketLen]uint64
+	Objects                 uint64
+	ObjSizes                sizeHistogram
+	Children                dataUsageHashMap
+	ReplicationPendingCount uint64
```
Do not add more entries here - this will cause dataUsageEntry to bloat and cause more memory usage. We should avoid it and keep it outside if not needed inside usage.

harshavardhana replied:
With this change, you also need to bump up the dataUsageEntry version such that it's updated otherwise it won't be read and it will clear off the old values.. look at

```go
// deserialize the supplied byte slice into the cache.
func (d *dataUsageCache) deserialize(r io.Reader) error {
        var b [1]byte
        n, _ := r.Read(b[:])
        if n != 1 {
                return io.ErrUnexpectedEOF
        }
```

harshavardhana replied:
We should also put this in a struct IMO such as 

```
type ReplicationStats struct {
     PendingSize
     FailedSize
     FailedCount
     PendingCount
     MissedThresholdSize
     AfterThresholdSize
     MissedThresholdCount
     AfterThresholdCount
}
```

We need all of this information and it looks like dataUsageEntry is not the place to calculate this - we have to come up with a better mechanism.

harshavardhana commented on cmd/bucket-replication-stats.go:
```diff
@@ -0,0 +1,142 @@
+	st := &ReplicationStats{
+		Cache: make(map[string]*BucketReplicationStats),
+	}
+	go func(st *ReplicationStats) {
```
Do not know why this is a goroutine @poornas - can this not be updated later? i.e you initialize it and then once the server is up and running you can ask globalReplicationStats.Refresh() to load from disk.

poornas replied:
this Cache needs to be set only once at the time of server initialization with values from the data usage cache.After that it accumulates in-memory stats - is there an issue with using go routine here, as it is just doing a one time load and then exiting the go routine?

harshavardhana replied:
Yeah but the problem is object layer is mostly not ready here. We should do it later when config is loaded and fully started.. 

We should do as a helper later to ensure that we get this value when server is fully up.

Commit:
Add API for replication metrics

Commit:
fix some of the calculations

harshavardhana commented on cmd/api-router.go:
```diff
@@ -311,7 +311,9 @@ func registerAPIRouter(router *mux.Router) {
-
+		// GetBucketReplicationStatus
+		router.Methods(http.MethodGet).HandlerFunc(
+			collectAPIStats("getbucketreplicationstatus", maxClients(httpTraceAll(api.GetBucketReplicationMetricsHandler)))).Queries("replication-metrics", "")
```
```suggestion
			collectAPIStats("getbucketreplicationmetrics", maxClients(httpTraceAll(api.GetBucketReplicationMetricsHandler)))).Queries("replication-metrics", "")
```

harshavardhana replied:
Can you move this all the way down to indicate these as extended API not part of AWS S3?

harshavardhana commented on cmd/bucket-replication-stats.go:
```diff
@@ -0,0 +1,143 @@
+	FailedSize uint64 `json:"objectsFailedReplicationTotalSize"`
+	// Total number of pending operations including metadata updates
+	OperationsPendingCount uint64 `json:"objectsPendingReplicationCount"`
+}
```
https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-metrics.html we should have at least two of these published by AWS S3 part of this. 

We should add how many objects didn't replicate in time? such as objects that missed SLA

harshavardhana commented on cmd/bucket-replication-stats.go:
```diff
@@ -0,0 +1,143 @@
+		if opType == replication.ObjectReplicationType {
+			atomic.AddUint64(&r.Cache[bucket].FailedSize, uint64(n))
+		}
+	case replication.Replica:
```
We also need replication types for these

```
        ObjectReplicationMissedThreshold
        ObjectReplicationReplicatedAfterThreshold
        ObjectReplicationNotTracked
```

harshavardhana committed:
non blocking queues replication

harshavardhana committed:
update versioning data-scanner

harshavardhana committed:
update last update structure change

harshavardhana committed:
more structure refactor

Commit:
Add MRF worker to handle spill-over replication operations

harshavardhana committed:
fix producer channel close

harshavardhana committed:
fix: multiple issues found with replication

- fixes an issue when client sends a bucket name
  with `/` at the end from SetRemoteTarget API call
  , make sure to trim the bucket name to avoid any
  extra `/` at the

- hold write locks in GetObjectNInfo during replication
  to ensure that object version stack is not overwritten
  while reading the content.

- add additional protection during WriteMetadata() to
  ensure that we always write a valid FileInfo{} and avoid
  ever writing empty FileInfo{} to the lowest layers.

Commit:
Fix cache update.
